core/state/state_object.go

```diff
@@ -286,10 +258,10 @@ func (s *stateObject) updateTrie() (Trie, error) {
 	var (
 		storage map[common.Hash][]byte
 		origin  map[common.Hash][]byte
+		hasher  = crypto.NewKeccakState()
 	)
-	tr, err := s.getTrie()
+	tr, err := s.storageTrie()
```

Comment:
So I guess the next step would be to replace with with `s.reader.ReadStorage` and `s.writer.WriteStorage` ?

Reply:
Unfortunately, trie is mandatory in any case for a full functional stateDB. As I said, I feel it's over abstracted by hiding it  to another interface.

But I do see the ugly integration code in verkle context, basically it would be nice if we can find a proper abstraction to handling trie loading difference between MPT and Verkle. For Verkle, it must use the same resolved trie for account and storage slot updates.

Reply:
but wait, so if we ever want to have a snapshot-only db, how would not abstracting the trie work?

Reply:
The snapshot-only db is only for reading purposes. But in case we want to execute the blocks and then **verify the result**, trie is mandatory to compute the state root for consensus comparison. 

For non-verification execution is something out of scope, let's forget it for now. It's why I say trie is mandatory except for read-only mode.